File: #15_3Sum/Solution.java
/* A O(n^2) solution */

import java.util.*;

public class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> result = new LinkedList<List<Integer>>();
        if (nums == null || nums.length < 3) return result;
        Arrays.sort(nums);
        for (int i = 0; i < nums.length - 2; i++) {
        	/* skip duplicates */
        	if (i > 0 && nums[i] == nums[i - 1]) continue;
        	int target = 0 - nums[i];
        	int lo = i + 1;
        	int hi = nums.length - 1;
        	while (lo < hi) {
        		if (nums[lo] + nums[hi] == target) {
        			List<Integer> newList = new LinkedList<Integer>();
        			newList.add(nums[i]);
					newList.add(nums[lo]);
					newList.add(nums[hi]);
					result.add(newList);
					/* skip duplicates */
					while (lo + 1 < nums.length && nums[lo + 1] == nums[lo]) lo++;
					while (hi - 1 > i && nums[hi - 1] == nums[hi]) hi--;
					hi -= 1; lo += 1;
        		}
        		else if (nums[lo] + nums[hi] > target) hi -= 1;
        		else lo += 1;
        	}
        }
        return result;
    }

    public static void main(String[] args) {
    	Solution sol = new Solution();
    	int[] t = {-1, 0, 1, 2, -1, -4};
    	List<List<Integer>> result = sol.threeSum(t);
    	for (List<Integer> ls : result) {
    		for (Integer it : ls) {
    			System.out.print(it + " ");
    		}
    		System.out.println();
    	}
    }
}

File: #15_3Sum/Solution.py
class Solution:
	# @param {integer[]} nums
	# @return {integer[][]}
	def threeSum(self, nums):
		if not nums or len(nums) < 3:
			return []
		re = []
		nums.sort()
		for x in range(0, len(nums) - 2):
			if x > 0 and nums[x] == nums[x - 1]:
				continue
			i = x + 1
			j = len(nums) - 1
			sum = 0 - nums[x]
			while i < j:
				if nums[i] + nums[j] > sum:
					j -= 1
					while i < j and nums[j] == nums[j + 1]:
						j -= 1
				elif nums[i] + nums[j] < sum:
					i += 1
					while i < j and nums[i] == nums[i - 1]:
						i += 1
				else:
					re.append([nums[x], nums[i], nums[j]])
					i += 1
					while i < j and nums[i] == nums[i - 1]:
						i += 1
					j -= 1
					while i < j and nums[j] == nums[j + 1]:
						j -= 1
		return re

sol = Solution()
s = [-2,0,1,1,2]
re = sol.threeSum(s)
for triple in re:
	for it in triple:
		print it,
	print